core/src/main/scala/pureconfig/DerivedWriters.scala

```diff
@@ -0,0 +1,113 @@
+package pureconfig
+
+import scala.collection.JavaConverters._
+import scala.collection.generic.CanBuildFrom
+import scala.language.higherKinds
+
+import com.typesafe.config._
+import pureconfig.error._
+import shapeless._
+import shapeless.labelled._
+
+/**
+ * Trait containing `ConfigWriter` instances for collection, product and coproduct types.
+ */
+trait DerivedWriters {
+
+  private[pureconfig] trait WrappedConfigWriter[Wrapped, SubRepr] extends ConfigWriter[SubRepr]
+
+  implicit final def hNilConfigWriter[Wrapped]: WrappedConfigWriter[Wrapped, HNil] = new WrappedConfigWriter[Wrapped, HNil] {
+    override def to(t: HNil): ConfigValue = ConfigFactory.parseMap(Map().asJava).root()
+  }
+
+  implicit final def hConsConfigWriter[Wrapped, K <: Symbol, V, T <: HList, U <: HList](
+    implicit
+    key: Witness.Aux[K],
+    vFieldConvert: Lazy[ConfigWriter[V]],
+    tConfigWriter: Lazy[WrappedConfigWriter[Wrapped, T]],
+    hint: ProductHint[Wrapped]): WrappedConfigWriter[Wrapped, FieldType[K, V] :: T] = new WrappedConfigWriter[Wrapped, FieldType[K, V] :: T] {
+
+    override def to(t: FieldType[K, V] :: T): ConfigValue = {
+      val keyStr = hint.configKey(key.value.toString().tail)
+      val rem = tConfigWriter.value.to(t.tail)
+      // TODO check that all keys are unique
+      vFieldConvert.value match {
+        case f: OptionConfigWriter[_] =>
+          f.toOption(t.head) match {
+            case Some(v) =>
+              rem.asInstanceOf[ConfigObject].withValue(keyStr, v)
+            case None =>
+              rem
+          }
+        case f =>
+          val fieldEntry = f.to(t.head)
+          rem.asInstanceOf[ConfigObject].withValue(keyStr, fieldEntry)
+      }
+    }
+  }
+
+  implicit final def cNilConfigWriter[Wrapped]: WrappedConfigWriter[Wrapped, CNil] = new WrappedConfigWriter[Wrapped, CNil] {
+    override def to(t: CNil): ConfigValue = ConfigFactory.parseMap(Map().asJava).root()
+  }
+
+  implicit final def coproductConfigWriter[Wrapped, Name <: Symbol, V, T <: Coproduct](
+    implicit
+    coproductHint: CoproductHint[Wrapped],
+    vName: Witness.Aux[Name],
+    vFieldConvert: Lazy[ConfigWriter[V]],
+    tConfigWriter: Lazy[WrappedConfigWriter[Wrapped, T]]): WrappedConfigWriter[Wrapped, FieldType[Name, V] :+: T] =
+    new WrappedConfigWriter[Wrapped, FieldType[Name, V] :+: T] {
+
+      override def to(t: FieldType[Name, V] :+: T): ConfigValue = t match {
+        case Inl(l) =>
+          // Writing a coproduct to a config can fail. Is it worth it to make `to` return a `Try`?
```

Comment:
I believe the comment is still relevant - it mentions a question for which there is still no clear answer. Though it may not be the best place to make TODOs and discussion topics visible, I don't mind having comments like this around for as long as it takes.